File: ghoststake-backend/internal/finance/rounds.go
package finance

import "math/big"

// Phase is a round as an observer sees it, which is not the same as the
// status the contract stores.
//
//	ParimutuelRound.phaseOf
//
// Two of these are functions of the clock rather than of storage: an OPEN
// round becomes CUTOFF the moment entry closes, without anything being
// written, and a LOCKED round sits in OBSERVATION until someone resolves it.
// Deriving them from stored status alone would show a stake button that is
// already dead, which is the bug GHO-18 fixed in the frontend.
type Phase string

const (
	PhaseOpen        Phase = "open"
	PhaseCutoff      Phase = "cutoff"
	PhaseObservation Phase = "observation"
	PhaseResolved    Phase = "resolved"
	PhaseVoid        Phase = "void"
)

// Stored statuses, as the projection records them.
const (
	StatusOpen     = "open"
	StatusLocked   = "locked"
	StatusResolved = "resolved"
	StatusVoid     = "void"
)

// PhaseOf derives the observable phase.
//
//	if resolved      -> Resolved
//	if void          -> Void
//	if locked        -> Observation
//	else             -> entry open ? Open : Cutoff
//
// `entryCutoff` is the contract's immutable: entry stops that many seconds
// before lockTime, so a pending lock transaction cannot be front-run by
// someone who can see it in the mempool.
func PhaseOf(status string, openTime, lockTime, entryCutoff, nowUnix int64) Phase {
	switch status {
	case StatusResolved:
		return PhaseResolved
	case StatusVoid:
		return PhaseVoid
	case StatusLocked:
		return PhaseObservation
	}
	if EntryIsOpen(status, openTime, lockTime, entryCutoff, nowUnix) {
		return PhaseOpen
	}
	return PhaseCutoff
}

// EntryIsOpen is whether a position can be taken right now.
//
//	ParimutuelRound.entryIsOpen:
//	  status == Open && now >= openTime && now < lockTime - entryCutoff
//
// Strictly less than the cutoff, matching the contract. An off-by-one in the
// permissive direction is a button that submits a transaction the contract
// will revert.
func EntryIsOpen(status string, openTime, lockTime, entryCutoff, nowUnix int64) bool {
	if status != StatusOpen {
		return false
	}
	return nowUnix >= openTime && nowUnix < lockTime-entryCutoff
}

// Odds is what one unit staked on a side returns if that side wins, WAD.
// 2e18 means "doubles your money".
//
//	ParimutuelRound.oddsOf:
//	  pool = up + down
//	  mulDiv(pool - mulDiv(pool, rake, WAD), WAD, sidePool)
//
// Provisional until lock: every subsequent entry on either side moves it. An
// empty side returns zero, which is not an infinite multiple but an undefined
// one — a side with nothing in it cannot win, because the minimum-side floor
// voids the round at lock.
func Odds(sidePool, upPool, downPool, rake *big.Int) *big.Int {
	if sidePool == nil || sidePool.Sign() == 0 {
		return new(big.Int)
	}
	pool := new(big.Int).Add(or0(upPool), or0(downPool))
	net := new(big.Int).Sub(pool, MulDiv(pool, or0(rake), WAD))
	return MulDiv(net, WAD, sidePool)
}

// Position is one user's stake in one round.
type Position struct {
	UpStake   *big.Int
	DownStake *big.Int
	// Claimed is whether the payout or refund has already been collected.
	Claimed bool
	// Leveraged is whether the stake was funded by borrowing: the router
	// opened the position, so the payout routes back through it to repay the
	// debt before anything reaches the user.
	Leveraged bool
}

// Claimable is what a user can collect from a round right now.
//
//	ParimutuelRound.claimableOf:
//	  if claimed            -> 0
//	  if void               -> up + down            (full refund, no rake)
//	  if not resolved       -> 0
//	  stake on winner == 0  -> 0
//	  else mulDiv(totalPool - rakeTaken, stake, winningPool)
//
// The floor division is the contract's, and so is its consequence: the sum of
// all payouts can fall a few wei short of the distributable pool. That dust
// stays in the contract rather than going to the last claimant, because a
// payout that depends on claim order is a payout people would race for.
//
// A void refunds the whole stake on both sides with no rake taken. That is why
// `rakeTaken` is always zero on a voided round — the protocol does not charge
// for a market that did not happen.
func Claimable(pos Position, status, winner string, upPool, downPool, rakeTaken *big.Int) *big.Int {
	if pos.Claimed {
		return new(big.Int)
	}
	if status == StatusVoid {
		return new(big.Int).Add(or0(pos.UpStake), or0(pos.DownStake))
	}
	if status != StatusResolved {
		return new(big.Int)
	}

	stake, winningPool := or0(pos.DownStake), or0(downPool)
	if winner == "up" {
		stake, winningPool = or0(pos.UpStake), or0(upPool)
	}
	if stake.Sign() == 0 || winningPool.Sign() == 0 {
		return new(big.Int)
	}

	total := new(big.Int).Add(or0(upPool), or0(downPool))
	distributable := new(big.Int).Sub(total, or0(rakeTaken))
	return MulDiv(distributable, stake, winningPool)
}

File: ghoststake-frontend/src/lib/rounds.ts
import { WAD } from "./format";

/** Mirrors `ParimutuelRound.Status`. */
export const Status = {
  None: 0,
  Open: 1,
  Locked: 2,
  Resolved: 3,
  Void: 4,
} as const;

/** Mirrors `ParimutuelRound.Phase`. */
export const Phase = {
  None: 0,
  Open: 1,
  Cutoff: 2,
  Observation: 3,
  Resolved: 4,
  Void: 5,
} as const;

export type PhaseValue = (typeof Phase)[keyof typeof Phase];

export const Side = { Up: 0, Down: 1 } as const;
export type SideValue = (typeof Side)[keyof typeof Side];

export type Round = {
  openTime: bigint;
  lockTime: bigint;
  closeTime: bigint;
  status: number;
  winner: number;
  lockPrice: bigint;
  closePrice: bigint;
  lockOracleRoundId: bigint;
  upPool: bigint;
  downPool: bigint;
  rakeTaken: bigint;
};

/**
 * Whether a new stake would actually be accepted right now.
 *
 * This is the entry cutoff, and it is the one piece of round state a UI can
 * get wrong in a way that costs a user money. `takePosition` refuses inside
 * `entryCutoff` seconds of the lock, so a button that stays live until
 * `lockTime` is lying for the last fifteen seconds — the user pays gas for a
 * transaction that was never going to land.
 *
 * The buffer exists to stop someone front-running the pending `lockRound`
 * transaction. Surfacing it honestly is the price of having it.
 */
export function entryOpen(round: Round, entryCutoff: bigint, nowSeconds: bigint): boolean {
  if (round.status !== Status.Open) return false;
  if (nowSeconds < round.openTime) return false;
  return nowSeconds + entryCutoff < round.lockTime;
}

/**
 * The instant entry actually closes, which is earlier than the lock.
 *
 * A countdown should run to this, not to `lockTime`.
 */
export function entryClosesAt(round: Round, entryCutoff: bigint): bigint {
  return round.lockTime - entryCutoff;
}

/**
 * What a side pays per unit staked, if it wins, at the pool's current shape.
 *
 * Derived, never stored: `(total - rake) / side`. It moves as people enter,
 * so it is a live quote and not a promise — which is the honest way to label
 * it in the UI too.
 *
 * Returns null while a side is empty, because the multiple is unbounded there
 * and rendering "∞" invites someone to read it as a payout.
 */
export function multipleFor(round: Round, side: SideValue, rake: bigint): bigint | null {
  const sidePool = side === Side.Up ? round.upPool : round.downPool;
  if (sidePool === 0n) return null;

  const total = round.upPool + round.downPool;
  const distributable = total - (total * rake) / WAD;
  return (distributable * WAD) / sidePool;
}

/**
 * Whether the round can still pay out at all.
 *
 * A side under `minSidePool` at lock voids the round and refunds everyone, so
 * a one-sided pool is worth warning about *before* someone joins the side
 * that is already ahead.
 */
export function willVoidOnLock(round: Round, minSidePool: bigint): boolean {
  return round.upPool < minSidePool || round.downPool < minSidePool;
}

export function phaseLabel(phase: PhaseValue): string {
  switch (phase) {
    case Phase.Open:
      return "Open";
    case Phase.Cutoff:
      return "Entry closed";
    case Phase.Observation:
      return "Running";
    case Phase.Resolved:
      return "Settled";
    case Phase.Void:
      return "Voided";
    default:
      return "Unknown";
  }
}

/** mm:ss, or h:mm:ss past an hour. Clamped at zero rather than going negative. */
export function formatCountdown(seconds: bigint): string {
  const total = seconds > 0n ? Number(seconds) : 0;
  const h = Math.floor(total / 3600);
  const m = Math.floor((total % 3600) / 60);
  const s = total % 60;
  const pad = (n: number) => n.toString().padStart(2, "0");
  return h > 0 ? `${h}:${pad(m)}:${pad(s)}` : `${pad(m)}:${pad(s)}`;
}
